File: leetcode/lc-us/cpp/904.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int totalFruit(vector<int> &tree) {
        unordered_map<int, int> map;
        int res = 0;
        for (int i = 0, j = 0; j < tree.size(); j++) {
            map[tree[j]]++;
            while (map.size() > 2) {
                map[tree[i]]--;
                if (!map[tree[i]]) map.erase(tree[i]);
                i++;
            }
            res = max(res, j - i + 1);
        }
        return res;
    }
};

File: leetcode/lc-us/py/904.py
from collections import defaultdict

from typing import List


class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        res = i = j = 0
        cnt = defaultdict(int)
        while j < len(fruits):
            cnt[fruits[j]] += 1
            while len(cnt) > 2:
                cnt[fruits[i]] -= 1
                if cnt[fruits[i]] == 0:
                    del cnt[fruits[i]]
                i += 1
            res = max(res, j - i + 1)
            j += 1
        return res
